### ISSUE TITLE
Retain timezone offset for datetime

### ISSUE THREAD

**Issue by gtsystem** [jcrist/msgspec#384]
### Description

It seems that msgspec transforms timezone aware timestamps to have the time represented in UTC. From the [doc](https://jcristharif.com/msgspec/supported-types.html#datetime):

_"During decoding any timezone-aware datetimes will have their timezone normalized to UTC."_

However this represents a loss of information as the original timezone offset is lost.

Is there any reason for this behavior? Any workaround or plan to address this limitation?

**Comment by jcrist** [comment]
The current behavior is intentional.

- Decoding into UTC uniformly allows for some performance optimizations, as we don't have to support allocating arbitrary timezone objects (which can be slow). We can reuse the UTC singleton object everywhere.
- In `MessagePack` all datetimes are encoded using UTC (there's no way to include an arbitrary offset). For uniformity across protocols, we implemented similar behavior in our JSON encoder/decoder
- In my experience, code handling timestamps generally should be written using UTC everywhere, rather than working with arbitrary timezones. Ensuring that all decoded timezone aware datetimes are already in UTC can simplify downstream consuming code.

That said, I'd be willing to reexamine this decision if given a compelling use case. Do you have a use case that requires keeping around the original timestamp? In what way are UTC datetimes insufficient?

**Comment by gtsystem** [comment]
I was thinking to adopt msgspec for https://github.com/gtsystem/lightkube, but today serialization and de-serialization is keeping the offset, so such change of behavior could be considered a breaking change. As the project is about operations, I'm 90% sure it should not break anything, but I cannot really rule out some use case where this is expected (Several operators are built on top of this library). 

Regarding use cases for date time with offset: One common use case is Transport & Logistics. For each point in a route we want to show the local time, but we also want to keep the offset in order to calculate the actual travel duration. This is also common for location services like POI search (show current time in that place, or opening times), as several POIs are showing in the map, but each of them may have a different offset. 

For reference, pydantic seems to retain the offset:
```python
from datetime import datetime
from pydantic import BaseModel

class Model(BaseModel):
    dt: datetime = None

m = Model(dt='2032-04-23T10:20:30.400+02:30')
print(m)
# Model(dt=datetime.datetime(2032, 4, 23, 10, 20, 30, 400000, tzinfo=datetime.timezone(datetime.timedelta(seconds=9000))))

print(m.json())
# '{"dt": "2032-04-23T10:20:30.400000+02:30"}'
```

And also encoding and decoding datetime from standard library seems to work as exected:
```python
>>> datetime.fromisoformat('2032-04-23T10:20:30.400000+02:30')
datetime.datetime(2032, 4, 23, 10, 20, 30, 400000, tzinfo=datetime.timezone(datetime.timedelta(seconds=9000)))
>>> datetime.fromisoformat('2032-04-23T10:20:30.400000+02:30').isoformat()
'2032-04-23T10:20:30.400000+02:30'
```

**Comment by jcrist** [comment]
Makes sense to me! We're in the middle of a cross-country move, but I still think I can get this resolved in the next week or two. Thanks for opening this issue, feedback like this is critical for improving the usefulness of `msgspec`.

**Comment by jcrist** [comment]
This has been fixed in #391. If you want to try it out before the next release, you can [install from github](https://jcristharif.com/msgspec/install.html#installing-from-github).

**Comment by gtsystem** [comment]
Wow, that was fast :) Thanks a lot!
